### 열 단위로 데이터 추출하기(32쪽)

- 데이터프레임에서 데이터를 열 단위로 추출하려면 대괄호와 열 이름을 사용해야 한다.
- 이때 열 이름은 꼭 작은 따옴표를 사용해서 지정해야 하고 추출한 열은 변수에 저장해서 사용할 수 있다.
- 이때 개의 열만 추출하면 시리즈를 얻을 수 있고 2개 이상의 열을 추출하면 데이터프레임을 얻을 수 있다.

- 다음은 데이터프레이(df)에서 열 이름이 country인 열을 추출하여 country_df에 저장한 것이다.
- type메서드를 사용하면 country_df에 저장된 데이터의 자료형이 시리즐라는 것을 확인할 수 있다.
- 시리즈도 head, tail 메서드를 가지고 있기 때문에 head, tail메서드로 가장 앞이나 뒤에 있는 5개의 데이터를 출력할 수 있다.

In [4]:
import pandas as pd

In [5]:
import pandas as pd
df = pd.read_csv('../data/gapminder.tsv', sep='\t')

In [6]:
country_df = df['country']

print(type(country_df))

<class 'pandas.core.series.Series'>


In [7]:
print(country_df.head())

0    Afghanistan
1    Afghanistan
2    Afghanistan
3    Afghanistan
4    Afghanistan
Name: country, dtype: object


In [9]:
print(country_df.tail())

1699    Zimbabwe
1700    Zimbabwe
1701    Zimbabwe
1702    Zimbabwe
1703    Zimbabwe
Name: country, dtype: object


- 리스트에 열 이름을 전달하면 여러 개의 열을 한 번에 추출할 수 있다.
- 다음은 이름이 country, continent, year인 열을 추출하여 변수 subset에 저장한 것이다.
- 이때 1개의 열이 아니라 2개 이상의 열을 추출했기 때문에 시리즈가 아니라 데이터프레임을 얻을 수 있다.

In [10]:
subset = df[['country', 'continent', 'year']]

print(type(subset))

<class 'pandas.core.frame.DataFrame'>


In [11]:
print(subset.head())

       country continent  year
0  Afghanistan      Asia  1952
1  Afghanistan      Asia  1957
2  Afghanistan      Asia  1962
3  Afghanistan      Asia  1967
4  Afghanistan      Asia  1972


In [12]:
print(subset.tail())

       country continent  year
1699  Zimbabwe    Africa  1987
1700  Zimbabwe    Africa  1992
1701  Zimbabwe    Africa  1997
1702  Zimbabwe    Africa  2002
1703  Zimbabwe    Africa  2007


## loc 속성으로 행 단위 데이터 추출하기(35쪽)

- 이번에는 데이터를 행 단위로 추출하는 방법에 대해 알아보자.
- 데이터를 행 단위로 추출하려면 loc, iloc속성을 사용해야 한다.

- loc속성에 대괄호를 이용하여 인덱스를 전달하면 행 데이터를 추출할 수 있다. 
- 다음은 인덱스가 0인 행 데이터를 추출한 것이다.
- 만약 -1과 같이 인덱스에 없는 값을 사용하면 오류가 발생하니 주의하자.

In [13]:
print(df.loc[0])

country      Afghanistan
continent           Asia
year                1952
lifeExp           28.801
pop              8425333
gdpPercap     779.445314
Name: 0, dtype: object


In [14]:
print(df.loc[99])

country      Bangladesh
continent          Asia
year               1967
lifeExp          43.453
pop            62821884
gdpPercap    721.186086
Name: 99, dtype: object


In [15]:
print(df.loc[-1])

KeyError: -1

- 만약 데이터프레임의 마지막 행 데이터를 추출하려면 어떻게 해야 할까? 마지막 데이터의 인덱스를 알아내야 한다.
- shaep[0]에 행 크기(1704)가 저장되어 있다는 점을 이용하여 마지막 행의 인덱스를 구하면 된다.
- 다음은 shape[0]에서 1을 뺀 값으로(1703) 마지막 행 데이터를 추출한 것이다.

In [20]:
number_of_rows = df.shape[0]
last_row_index = number_of_rows - 1

print(df.loc[last_row_index])

country        Zimbabwe
continent        Africa
year               2007
lifeExp          43.487
pop            12311143
gdpPercap    469.709298
Name: 1703, dtype: object


- 데이터프레임의 마지막 행 데이터를 추출하는 또 다른 하나는 tail메서드를 사용하는 것이다.
- 다음과 같이 tail 메서드의 인자 n에 1을 전달하면 마지막 행의 데이터를 추출할 수 있다.

In [22]:
print(df.tail(n=1))

       country continent  year  lifeExp       pop   gdpPercap
1703  Zimbabwe    Africa  2007   43.487  12311143  469.709298


In [23]:
print(df.tail(n=2))

       country continent  year  lifeExp       pop   gdpPercap
1702  Zimbabwe    Africa  2002   39.989  11926563  672.038623
1703  Zimbabwe    Africa  2007   43.487  12311143  469.709298


- 만약 인덱스가 0, 99, 999인 데이터를 한 번에 추출하려면 리스트에 원하는 인덱스를 담아 loc속성에 전달하면 된다.

In [24]:
print(df.loc[[0, 99, 999]])

         country continent  year  lifeExp       pop    gdpPercap
0    Afghanistan      Asia  1952   28.801   8425333   779.445314
99    Bangladesh      Asia  1967   43.453  62821884   721.186086
999     Mongolia      Asia  1967   51.253   1149500  1226.041130


## tail 메서드와 loc 속성이 반환하는 자료형은 서로 달라요!(37쪽)

- tail 메서드와 loc속성이 반환하는 데이터의 자료형은 다르다.
- 다음은 tail 메서드와 loc속성으로 추출한 데이터의 자료형을 type메서드로 확인한 것이다.
- loc속성이 반환한 데이터 자료형은 시리즈이고 tail메서드가 반환한 데이터 자료형은 데이터프레임이다.

In [25]:
subset_loc = df.loc[0] 
subset_tail = df.tail(n=1)

print(type(subset_loc))
print(type(subset_tail))

<class 'pandas.core.series.Series'>
<class 'pandas.core.frame.DataFrame'>


# 직접 해보세요!
## iloc 속성으로 행 데이터 추출하기(37쪽)

- 이번에는 iloc속성으로 행 데이터를 추출하는 방법에 대해 알아보겠다.
- loc속성은 데이터프레임의 인덱스를 사용하여 데이터를 추출했지만 iloc속성은 데이터 순서를 의미하는 행 번호를 사용하여 데이터를 추출한다.
- 지금은 인덱스와 행 번호가 동일하여 동일한 결과값을 출력한다. 
- 다음은 iloc 속성에 1을 전달하여 데이터를 추출한 것이다.

In [27]:
print(df.iloc[1])

country      Afghanistan
continent           Asia
year                1957
lifeExp           30.332
pop              9240934
gdpPercap      820.85303
Name: 1, dtype: object


In [28]:
print(df.iloc[99])

country      Bangladesh
continent          Asia
year               1967
lifeExp          43.453
pop            62821884
gdpPercap    721.186086
Name: 99, dtype: object


- iloc 속성은 음수를 사용해도 데이터를 추출할 수 있다.
- 다음은 -1을 전달하여 마지막 행 데이터를 추출한 것이다. 하지만 데이터프레임이 아예 존재하지 않는 행 번호를 전달하면 오류가 발생한다.

In [29]:
print(df.iloc[-1])

country        Zimbabwe
continent        Africa
year               2007
lifeExp          43.487
pop            12311143
gdpPercap    469.709298
Name: 1703, dtype: object


In [30]:
print(df.iloc[1710])

IndexError: single positional indexer is out-of-bounds

- iloc속성도 여러 데이터를 한 번에 추출할 수 있다. loc속성을 사용했던 것 처럼 원하는 데이터의 행 번호를 리스트에 담아 전달하면 된다.

In [31]:
print(df.iloc[[0, 99, 999]])

         country continent  year  lifeExp       pop    gdpPercap
0    Afghanistan      Asia  1952   28.801   8425333   779.445314
99    Bangladesh      Asia  1967   43.453  62821884   721.186086
999     Mongolia      Asia  1967   51.253   1149500  1226.041130


# 직접 해보세요!
## 데이터 추출하기 ─ 슬라이싱 구문, range 메서드(39쪽)
### 1. 슬라이싱 구문으로 데이터 추출하기

- 다음은 모든 행(:)의 데이터에 대해 year, pop열을 추출하는 방법이다. 이때 loc와 iloc 속성에 전달하는 열 지정값은 반드시 형식에 맞게 전달해야 한다.
- 예를 들어 loc속성의 열 지정값에 정수 리스트를 전달하면 오류가 발생한다.
- 반대로 iloc속성의 열 지정 값에 문자열 리스트를 전달하면 오류가 발생한다.

In [32]:
subset = df.loc[:, ['year', 'pop']] 
print(subset.head())

   year       pop
0  1952   8425333
1  1957   9240934
2  1962  10267083
3  1967  11537966
4  1972  13079460


In [33]:
subset = df.iloc[:, [2, 4, -1]] 
print(subset.head())

   year       pop   gdpPercap
0  1952   8425333  779.445314
1  1957   9240934  820.853030
2  1962  10267083  853.100710
3  1967  11537966  836.197138
4  1972  13079460  739.981106


### 2. range 메서드로 원하는 데이터 추출하기

- 이번에는 iloc속성과 파이썬 내장 메서드인 range를 응용하는 방법에 대해 알아보자.
- range메서드는 지정한 구간의 정수 리스트를 반환한다. iloc속성의 열 지정값에는 정수 리스트를 전달해야 한다는 점과 range메서드의 반환값이 정수 리스트인 점을 이용하여 원하는 데이터를 추출하는 것이다.
- range메서드는 지정한 범위의 정수 리스트를 반환하는 것이 아니라 제네레이터를 반환한다.
- iloc 속성은 제너레이터로 데이터 추출을 할 수 없다. 다행이 제너레이터는 간단하게 리스트로 변환할 수 있다.
- 다음은 range(5)가 반환한 제너레이터를 정수값을 가진 리스트 [0,1,2,3,4]로 변환하여 iloc의 열 지정 값에 전달한 것이다. 자주 사용하는 방법은 아니지만 알아두면 유용하다.

In [34]:
small_range = list(range(5)) 
print(small_range)

[0, 1, 2, 3, 4]


In [35]:
print(type(small_range))

<class 'list'>


In [36]:
subset = df.iloc[:, small_range] 
print(subset.head())

       country continent  year  lifeExp       pop
0  Afghanistan      Asia  1952   28.801   8425333
1  Afghanistan      Asia  1957   30.332   9240934
2  Afghanistan      Asia  1962   31.997  10267083
3  Afghanistan      Asia  1967   34.020  11537966
4  Afghanistan      Asia  1972   36.088  13079460


In [37]:
small_range = list(range(3, 6)) 
print(small_range)

[3, 4, 5]


In [38]:
subset = df.iloc[:, small_range] 
print(subset.head())

   lifeExp       pop   gdpPercap
0   28.801   8425333  779.445314
1   30.332   9240934  820.853030
2   31.997  10267083  853.100710
3   34.020  11537966  836.197138
4   36.088  13079460  739.981106


- range메서드에 range(0,6,2)와 같은 방법으로 3개의 인자를 전달하면 0부터 5까지 2만큼 건너뛰는 제네레이터를 생성한다.
- 이 제네레이터를 리스트로 변환하면 범위는 0~5이고 짝수로 된 정수 리스트를 얻을 수 있다.

In [36]:
small_range = list(range(0, 6, 2)) 
subset = df.iloc[:, small_range] 
print(subset.head())

       country  year       pop
0  Afghanistan  1952   8425333
1  Afghanistan  1957   9240934
2  Afghanistan  1962  10267083
3  Afghanistan  1967  11537966
4  Afghanistan  1972  13079460


### 4. 슬라이싱, range 메서드 비교하기

- 실무에서는 range메서드보다 간편하게 사용할 수 있는 파이썬 슬라이싱 구문을 선호한다.
- range메서드가 변환한 제너레이터를 리스트로 변환하는 과정을 거치지 않아도 되기 때문이다.
- 예를 들어 list(range(3))과 [:3]의 결과값은 동일한다.

In [37]:
subset = df.iloc[:, :3] 
print(subset.head())

       country continent  year
0  Afghanistan      Asia  1952
1  Afghanistan      Asia  1957
2  Afghanistan      Asia  1962
3  Afghanistan      Asia  1967
4  Afghanistan      Asia  1972


- 0:6:2를 열 지정값에 전달하면 3에서 얻은 결과값과 동일한 결과값을 얻을 수 있다.

In [38]:
subset = df.iloc[:, 0:6:2] 
print(subset.head())

       country  year       pop
0  Afghanistan  1952   8425333
1  Afghanistan  1957   9240934
2  Afghanistan  1962  10267083
3  Afghanistan  1967  11537966
4  Afghanistan  1972  13079460


### 5. loc, iloc 속성 자유자재로 사용하기

- 만약 loc속성으로 0, 99, 999번째 행의 0, 3, 5번째 열 데이터를 추출하려면 다음과 같이 코드를 작성하면 된다.

In [39]:
print(df.iloc[[0, 99, 999], [0, 3, 5]])

         country  lifeExp    gdpPercap
0    Afghanistan   28.801   779.445314
99    Bangladesh   43.453   721.186086
999     Mongolia   51.253  1226.041130


- iloc속성의 열 지정값으로 정수리스트를 전달하는 것이 간편해보일 수 있으나, 이렇게 작성한 코드는 나중에 어떤 데이터를 추출하기 위한 코드인지 파악하기 어렵다.
- 그래서 보통은 다음과 같은 방법으로 loc속성을 이용하여 열 지정값으로 열 이름을 전달한다.

In [40]:
print(df.loc[[0, 99, 999], ['country', 'lifeExp', 'gdpPercap']])

         country  lifeExp    gdpPercap
0    Afghanistan   28.801   779.445314
99    Bangladesh   43.453   721.186086
999     Mongolia   51.253  1226.041130


In [41]:
print(df.loc[10:13, ['country', 'lifeExp', 'gdpPercap']])

        country  lifeExp    gdpPercap
10  Afghanistan   42.129   726.734055
11  Afghanistan   43.828   974.580338
12      Albania   55.230  1601.056136
13      Albania   59.280  1942.284244
